pkg/dio/writer.go

```diff
@@ -32,65 +32,51 @@ type DynamicMultiWriter interface {
 }
 
 type multiWriter struct {
-	mutex     sync.Mutex
-	waitGroup sync.WaitGroup
-
+	mutex   sync.Mutex
 	writers []io.Writer
 }
 
 func (t *multiWriter) Write(p []byte) (int, error) {
 	var n int
 	var err error
 
-	var wTmp []io.Writer
-	if verbose {
-		defer func() {
-			log.Debugf("[%p] write %q to %d writers (err: %#+v)", t, string(p[:n]), len(wTmp), err)
-		}()
-	}
-
 	t.mutex.Lock()
-
-	t.waitGroup.Add(1)
-	defer t.waitGroup.Done()
-
-	// stash a local copy of the slice as we never want to write twice to a single writer
-	// if remove is called during this flow
-	wTmp = make([]io.Writer, len(t.writers))
-	copy(wTmp, t.writers)
-
-	t.mutex.Unlock()
+	defer t.mutex.Unlock()
 
 	eof := 0
 	// possibly want to add buffering or parallelize this
-	for _, w := range wTmp {
+	for _, w := range t.writers {
 		n, err = w.Write(p)
 		if err != nil {
 			// remove the writer
 			log.Debugf("[%p] removing writer %p due to %s", t, w, err.Error())
-
 			// Remove grabs the lock
-			t.Remove(w)
-
+			t.remove(w)
```

Comment:
remove no longer grabs the lock - please make sure you update comments if you're changing behaviour as this gets very, very confusing.